File: AndroidBase/androidBase/src/main/java/ams/android_base/utils/RuntimePermissionsManger.java
package ams.android_base.utils;

import android.content.Context;

import com.yanzhenjie.permission.AndPermission;
import com.yanzhenjie.permission.Options;
import com.yanzhenjie.permission.Rationale;
import com.yanzhenjie.permission.RequestExecutor;

import java.util.List;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;

@SuppressWarnings("unused")
public class RuntimePermissionsManger {

    @SuppressWarnings("FieldCanBeLocal")
    private AppCompatActivity appCompatActivity;
    @SuppressWarnings("FieldCanBeLocal")
    private Fragment fragment;
    private String okButtonTitle;
    private String cancelButtonTitle;
    private String permissionDialogTitle;
    private String permissionDescription;
    private Boolean enableRationalMessage = false;

    public RuntimePermissionsManger (AppCompatActivity appCompatActivity) { this.appCompatActivity = appCompatActivity; }
    public RuntimePermissionsManger (Fragment fragment) { this.fragment = fragment; }

    public void setOkButtonTitle(String okButtonTitle) { this.okButtonTitle = okButtonTitle; }
    public void setCancelButtonTitle(String cancelButtonTitle) { this.cancelButtonTitle = cancelButtonTitle; }
    public void setPermissionDescription(String permissionDescription) { this.permissionDescription = permissionDescription; }
    public void setPermissionDialogTitle(String permissionDialogTitle) { this.permissionDialogTitle = permissionDialogTitle; }
    public void setEnableRationalMessage(Boolean enableRationalMessage) { this.enableRationalMessage = enableRationalMessage; }

    private AlertDialog.Builder customDialog (Context context, final RequestExecutor executor) {
        AlertDialog.Builder dialog = new AlertDialog.Builder(context);
        dialog.setCancelable(true);
        if (permissionDialogTitle != null) dialog.setTitle(permissionDialogTitle);
        // List<String> permissionNames = Permission.transformText(context, permissions);
        // String message = permissionDescription + "\n" + TextUtils.join("\n", permissionNames);
        dialog.setMessage(permissionDescription);
        dialog.setPositiveButton(okButtonTitle, (dialog12, which) -> executor.execute());
        dialog.setNegativeButton(cancelButtonTitle, (dialog1, which) -> executor.cancel());
        return dialog;
    }

    private Rationale customRationale() {
        if (okButtonTitle != null && cancelButtonTitle != null && permissionDescription != null && enableRationalMessage) {
            return (context, data, executor) -> customDialog(context, executor).show();
        }
        return null;
    }

    private Options getPermissionsRequestOptions () {
        Options andPermissionOptions;
        if (appCompatActivity != null) andPermissionOptions = AndPermission.with(appCompatActivity);
        else andPermissionOptions = AndPermission.with(appCompatActivity);
        return andPermissionOptions;
    }

    public void requestSinglePermission (String permission, final RuntimePermissionsListener listener) {
        getPermissionsRequestOptions()
                .runtime()
                .permission(permission)
                .rationale(customRationale())
                .onGranted(grantedPermission -> listener.onPermissionGranted())
                .onDenied(deniedPermission -> listener.onPermissionDenied(permission))
                .start();
    }

    public void requestPermissions (String[] permissions, final RuntimePermissionsListener listener) {
        getPermissionsRequestOptions()
                .runtime()
                .permission(permissions)
                .rationale(customRationale())
                .onGranted(grantedPermissions -> {
                    listener.onPermissionsGranted();
                    listener.onPermissionsDenied(grantedPermissions);
                })
                .onDenied(listener::onPermissionsDenied)
                .start();
    }

    // Listeners
    public interface RuntimePermissionsListener {
        default void onPermissionsGranted() {}
        default void onPermissionsGranted(List<String> grantedPermissions) {}
        default void onPermissionsDenied(List<String> deniedPermissions) {}
        default void onPermissionGranted() {}
        default void onPermissionDenied(String deniedPermission) {}
    }


    /* This is how to request one or many permissions (Do not forget to add the permissions in the manifest)
     String[] PERMISSIONS = {Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.CAMERA};
        RuntimePermissionsManger manger = new RuntimePermissionsManger((AppCompatActivity) context);
        manger.setCancelButtonTitle("No");
        manger.setOkButtonTitle("Yes");
        manger.setPermissionDescription("Message");
        manger.setEnableRationalMessage(true);
        manger.requestPermissions(PERMISSIONS, new RuntimePermissionsManger.RuntimePermissionsListener() {
            @Override
            public void onPermissionsGranted() {
            }
            @Override
            public void onPermissionsDenied(String[] deniedPermissions) {
            }
        });
    */

}

File: androiddevkit/src/main/java/com/ams/androiddevkit/utils/RuntimePermissionsManger.kt
@file:Suppress("unused")

package com.ams.androiddevkit.utils

import android.content.Context
import android.content.pm.PackageManager
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.fragment.app.Fragment
import com.ams.androiddevkit.utils.runtimePermissions.BaseRationale
import com.yanzhenjie.permission.AndPermission
import com.yanzhenjie.permission.option.Option
import com.yanzhenjie.permission.runtime.PermissionRequest

@Suppress("MemberVisibilityCanBePrivate")
open class RuntimePermissionsManger {

    protected var appCompatActivity: AppCompatActivity? = null
    protected var fragment: Fragment? = null
    protected var okButtonTitle: String? = null
    protected var cancelButtonTitle: String? = null
    protected var permissionDialogTitle: String? = null
    protected var permissionDescription: String? = null
    protected var enableRationalMessage = false
    protected val permissionsRequestOption: Option by lazy {
        if (appCompatActivity != null) AndPermission.with(appCompatActivity) else AndPermission.with(fragment)
    }

    constructor(appCompatActivity: AppCompatActivity) { this.appCompatActivity = appCompatActivity }

    constructor(fragment: Fragment) { this.fragment = fragment }

    fun setOkButtonTitle(okButtonTitle: String): RuntimePermissionsManger {
        this.okButtonTitle = okButtonTitle
        return this
    }

    fun setCancelButtonTitle(cancelButtonTitle: String): RuntimePermissionsManger {
        this.cancelButtonTitle = cancelButtonTitle
        return this
    }

    fun setPermissionDialogTitle(permissionDialogTitle: String): RuntimePermissionsManger {
        this.permissionDialogTitle = permissionDialogTitle
        return this
    }

    fun setPermissionDescription(permissionDescription: String): RuntimePermissionsManger {
        this.permissionDescription = permissionDescription
        return this
    }

    fun setEnableRationalMessage(enableRationalMessage: Boolean): RuntimePermissionsManger {
        this.enableRationalMessage = enableRationalMessage
        return this
    }

    fun withRational(enableRationalMessage: Boolean): RuntimePermissionsManger {
        this.enableRationalMessage = enableRationalMessage
        return this
    }

    open fun getCustomRationale(): BaseRationale<List<String>> {
        return BaseRationale<List<String>>()
            .setOkButtonTitle(okButtonTitle)
            .setCancelButtonTitle(cancelButtonTitle)
            .setPermissionDialogTitle(permissionDialogTitle)
            .setPermissionDescription(permissionDescription)
    }

    open fun getRunTimePermissionsBuilder(vararg permissions: String): PermissionRequest {
        val builder = permissionsRequestOption
            .runtime()
            .permission(permissions)
        if (enableRationalMessage) {
            builder.rationale(getCustomRationale())
        }
        return builder
    }

    open fun requestSinglePermission(permission: String, listener: RuntimePermissionsListener) {
        getRunTimePermissionsBuilder(permission)
            .onGranted { listener.onPermissionGranted() }
            .onDenied { listener.onPermissionDenied(permission) }
            .start()
    }

    open fun requestPermissions(vararg permissions: String, listener: RuntimePermissionsListener) {
        getRunTimePermissionsBuilder(*permissions)
            .onGranted { grantedPermissions: List<String> ->
                listener.onPermissionsGranted()
                listener.onPermissionsGranted(grantedPermissions)
            }
            .onDenied { deniedPermissions: List<String> -> listener.onPermissionsDenied(deniedPermissions) }
            .start()
    }

    open fun isPermissionGranted(context: Context, permission: String): Boolean {
        return ActivityCompat.checkSelfPermission(context, permission) == PackageManager.PERMISSION_GRANTED
    }

    // Listeners
    interface RuntimePermissionsListener {
        fun onPermissionsGranted() {}
        fun onPermissionsGranted(grantedPermissions: List<String?>?) {}
        fun onPermissionsDenied(deniedPermissions: List<String?>?) {}
        fun onPermissionGranted() {}
        fun onPermissionDenied(deniedPermission: String?) {}
    }

    /* This is how to request one or many permissions (Do not forget to add the permissions in the manifest)
     String[] PERMISSIONS = {Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.CAMERA};
        RuntimePermissionsManger manger = new RuntimePermissionsManger((AppCompatActivity) context);
        manger.setCancelButtonTitle("No");
        manger.setOkButtonTitle("Yes");
        manger.setPermissionDescription("Message");
        manger.setEnableRationalMessage(true);
        manger.requestPermissions(PERMISSIONS, new RuntimePermissionsManger.RuntimePermissionsListener() {
            @Override
            public void onPermissionsGranted() {
            }
            @Override
            public void onPermissionsDenied(String[] deniedPermissions) {
            }
        });
    */
}